esda/correlogram.py

```diff
@@ -0,0 +1,153 @@
+    how the relationship between spatial units changes over different notions of scale.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        geodataframe holding spatial and attribute data
+    variable: str
+        column on the geodataframe used to compute autocorrelation statistic
+    distances : list
+        list of distances to compute the autocorrelation statistic
+    distance_type : str, optional
+        which concept of distance to increment. Options are {`band`, `knn`}.
+        by default 'band' (for `libpysal.weights.DistanceBand` weights)
+    statistic : str, by default 'I'
+        which spatial autocorrelation statistic to compute. Options in {`I`, `G`, `C`}
+    weights_kwargs : dict
+        additional keyword arguments passed to the libpysal.weights.W class
+    stat_kwargs : dict
+        additional keyword arguments passed to the `esda` autocorrelation statistic class.
+        For example for faster results with no statistical inference, set the number
+        of permutations to zero with {permutations: 0}
+    n_jobs : int
+        number of jobs to pass to joblib. If -1 (default), all cores will be used
+    backend : str
+        backend parameter passed to joblib
+
+    Returns
+    -------
+    outputs : pandas.DataFrame
+        table of autocorrelation statistics at increasing distance bandwidths
+    """
+    if stat_kwargs is None:
+        stat_kwargs = dict()
+    if weights_kwargs is None:
+        weights_kwargs = dict()
+    if statistic == "I":
+        STATISTIC = Moran
+    elif statistic == "G":
+        STATISTIC = G
+    elif statistic == "C":
+        STATISTIC = Geary
+    else:
+        with NotImplementedError as e:
+            raise e("Only I, G, and C statistics are currently implemented")
+
+    if distance_type == "band":
+        W = DistanceBand
+    elif distance_type == "knn":
+        if max(distances) > gdf.shape[0] - 1:
+            with ValueError as e:
+                raise e("max number of neighbors must be less than or equal to n-1")
+        W = KNN
+    else:
+        with NotImplementedError as e:
+            raise e("distance_type must be either `band` or `knn` ")
+
+    #  should be able to build the tree once and reuse it?
+    #  but in practice, im not seeing any real difference from starting a new W from scratch each time
+    pts = get_points_array(gdf[gdf.geometry.name])
+    tree = KDTree(pts)
+    y = gdf[variable].values
+
+    inputs = [
+        tuple(
+            [
+                y,
+                tree,
+                W,
+                STATISTIC,
+                dist,
+                weights_kwargs,
+                stat_kwargs,
+            ]
+        )
+        for dist in distances
+    ]
+
+    outputs = Parallel(n_jobs=n_jobs, backend=backend)(
+        delayed(_get_autocorrelation_stat)(i) for i in inputs
+    )
+
+    return (
+        pd.DataFrame(outputs)
+        .select_dtypes(["number"])
```

Comment:
so, in the abstract version, this function becomes `spatial_profile(callable, y, gdf, **kwargs)`, where `callable` is any function that takes (y,w) as arguments.

in that case, should we get rid of this `.select_dtypes` here, and return everything the callable has? Also, since [this line](https://github.com/knaaptime/esda/blob/f2b952e54933d17fd9d88ad82838f3c45d40b06a/esda/correlogram.py#L41) requires a *class*, not a generic callable.

Reply:
I think just keep it as correlogram. Allowing a user defined function already means we can't assume/ensure the output is "correlation", but it's a relatively advanced way to work with the function.

And yes, the output type of the callable has to be a namespace/namedtuple/object kind of thing, but I think that's an ok requirement